### PR TITLE
submission: resource_dasm and phosg

### PR DESCRIPTION
**Pull request by fracai**
#### Description

Portfiles for `resource_dasm` and dependency `phosg`

`phosg` doesn't have any releases and uses the latest commit (as of submission).
`resource_dasm` does have some release tags, but there have been many commits since the last tag, so also uses the latest commit (as of submission).

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 13.2.1 22D68 arm64
Xcode 14.2 14C18

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: devel/phosg/Portfile
Comment (harens) [inline devel/phosg/Portfile:17]: This is set by the github portgroup

File: devel/phosg/Portfile
Comment (harens) [inline devel/phosg/Portfile:23]: Could we use Python 3.11?

**Review by harens — COMMENTED** [review]
Hi there @fracai 👋 Thank you for your first contribution to MacPorts! You've done a great job here. I've left a few minor comments on phosg, which also apply to resource_dasm.

If you could separate the two portfiles into different commits in the form `NAME: new port`, that would be great. Thank you for contributing @fracai 👍

**Comment by fracai** [comment]
> If you could separate the two portfiles into different commits in the form `NAME: new port`, that would be great.

Do you mean that I should create two PRs? Or do you want me to force push (or open a new replacement PR) with one commit for each Portfile, with `NAME: new port` as the commit message?

Thanks for the review, I'll get the changes up once I understand the commit format you're looking for.

File: devel/phosg/Portfile
Comment (fracai) [reply devel/phosg/Portfile:8]: I originally set a revision because I was only setting a commit hash to checkout. If I now set the version number as the commit date, should I hold off on setting a revision until it's needed?

Comment (harens) [reply devel/phosg/Portfile:8]: > If I now set the version number as the commit date, should I hold off on setting a revision until it's needed?

Setting the revision number explicitly as zero should be good.

Comment (fracai) [reply devel/phosg/Portfile]: `revision 0` in both

File: devel/phosg/Portfile
Comment (fracai) [reply devel/phosg/Portfile:7]: `version` dated in both

File: devel/phosg/Portfile
Comment (fracai) [reply devel/phosg/Portfile]: removed

File: devel/phosg/Portfile
Comment (fracai) [reply devel/phosg/Portfile]: removed

File: devel/phosg/Portfile
Comment (fracai) [reply devel/phosg/Portfile]: `pyver` set to 3.11

**Comment by fracai** [comment]
comments resolved, split to two commits, and force pushed

**Comment by harens** [comment]
```
   CMake Error at CMakeLists.txt:30 (find_package):
    By not providing "Findphosg.cmake" in CMAKE_MODULE_PATH this project has
    asked CMake to find a package configuration file provided by "phosg", but
    CMake did not find one.
  
    Could not find a package configuration file provided by "phosg" with any of
    the following names:
  
      phosgConfig.cmake
      phosg-config.cmake
  
    Add the installation prefix of "phosg" to CMAKE_PREFIX_PATH or set
    "phosg_DIR" to a directory containing one of the above files.  If "phosg"
    provides a separate development package or SDK, be sure it has been
    installed.
```

**Comment by fracai** [comment]
Gah, I forgot to include cmake as a dependency.
I actually figured it would have been included as a dependency as it's part of the cmake portgroup.

**Comment by harens** [comment]
> I actually figured it would have been included as a dependency as it's part of the cmake portgroup.

It is a dependency, so that's not the problem:

https://github.com/macports/macports-ports/blob/fd259e6117af72e16caa5c8e85b4aa5ad3d532ca/_resources/port1.0/group/cmake-1.1.tcl#L91-L93

The issue is that `phosg-config.cmake` (as well as `phosg-debug.cmake` and `phosg.cmake`) are installed into `${prefix}/lib`, when they should be in their own subdirectory. Maybe `${prefix}/lib/cmake/phosg/`?

**Comment by fracai** [comment]
The latest workflow failed `resource_dasm` because `cmake` couldn't find `phosg`. It suggests adding the `phosg` installation prefix, but the `Portfile` already does that with `configure.args-append … -DCMAKE_PREFIX_PATH=${prefix}/lib`. And when building the Portfile, adding this arg did get the build working. Does the workflow use trace mode ([currently broken on Ventura](https://trac.macports.org/ticket/66358)) and that's causing an issue?
I note that line [1304 of the log](https://github.com/macports/macports-ports/actions/runs/4495960166/jobs/7919345614#step:8:1310) has two occurrences of `DCMAKE_PREFIX_PATH`. I wonder if only the first is taking effect.
The build also suggests setting `phosg_DIR`, so I'll give that a try instead of `DCMAKE_PREFIX_PATH`.

**Comment by fracai** [comment]
Hah, didn't see your comment while posting mine.
So I need to change the installation of just those `*.cmake` files to a phosg subdir? And then update the `resource_dasm` argument accordingly?
And then remove `cmake` as a dependency as well?

**Comment by fracai** [comment]
The latest push installs the cmake files to `lib/cmake/<port>` and everything seems to compile without the extra cmake arguments. I'll be interested to see if it works with the workflow builds.

I just saw your additional notes, so I haven't addressed those yet. I'll take a look.

**Comment by fracai** [comment]
So, my initial pass sums up to 🤷 

It definitely compiles and is useable by resource_dasm as I've compiled both and successfully tested resource_dasm while phosg was installed under /opt/local. So I'm tempted to say that issue might be down to how that individual built, installed, and linked to phosg. And the C++20 request is perhaps not actually using features introduced in 20 as, again, it did build for me?

I don't know. If the workflow build succeeds I'm tempted to keep an eye on those issues and make updates accordingly, but maybe not let them become blockers… ?

**Comment by fracai** [comment]
This latest sets `phosg_DIR`
🤞

**Comment by fracai** [comment]
I'm at a loss. The arg `DCMAKE_PREFIX_PATH` already exists with a different path, so setting it in the Portfile adds a second copy, but doesn't build on the workflow. Setting `Dphosg_DIR` also doesn't build on the workflow. Both of these build for me locally.

I'm assuming the workflow uses tracemode, which I can't test with as it's broken on Ventura.

Is it possible that there's some other difference how the workflow builder operates? Or something I can do to test in a different way?

File: devel/phosg/Portfile
Comment (harens) [inline devel/phosg/Portfile:5]: Would you mind adding the following (or something similar) to both this and the other port? (see [wiki](https://guide.macports.org/#reference.portgroup.github.distfilestrategy))

```tcl
github.tarball_from archive
```

By default, the tar is fetched from `https://codeload.github.com/fuzziqersoftware/phosg/legacy.tar.gz/e1bd000?dummy=`, which is a legacy method.

Note that this would then require the checksums to be changed. Run `sudo port clean --all NAME` to remove the old downloaded files.

**Review by harens — COMMENTED** [review]
Ok I don't know how I didn't spot this before, but phosg hasn't been set as a dependency of resource_dasm. That's likely the issue.

The C++20 thing affects both ports, and might prevent them from building on older hardware ([wiki](https://trac.macports.org/wiki/CompilerSelection)). Maybe take a look at https://github.com/macports/macports-ports/pull/9932#issuecomment-774422453 or https://github.com/macports/macports-ports/pull/14963#issuecomment-1140330450.

I think after that we can merge this and see what happens with the buildbots. Thanks for all your work on this @fracai :)

**Comment by fracai** [comment]
How did I manage to not specify `phosg` as a dependency. Well, I know exactly how. I'm testing locally without trace so I installed `phosg` and then `resource_dasm` and the dependency was met. But still, how did I forget to specify that?

Oh well.

This latest adds the missing dependency and the compiler blacklist.

Thanks for your help in figuring this out.

File: devel/resource_dasm/Portfile
Comment (harens) [inline devel/resource_dasm/Portfile:31]: phosg should probably be a library dependency (i.e. required at both build time and runtime). I'm guessing the same is true for zlib as well.

File: devel/resource_dasm/Portfile
Comment (harens) [inline devel/resource_dasm/Portfile:33]: [Upstream docs](https://github.com/fuzziqersoftware/resource_dasm#building) note that "Also, this [Netpbm] is a runtime dependency only; you can install it later if you find that you need it, and you won't have to rebuild resource_dasm."

That can then be a runtime rather than build time dependency.

File: devel/resource_dasm/Portfile
Comment (harens) [inline devel/resource_dasm/Portfile:11]: This might need some tweaking depending on what C++20 features (if any) are used, but we'll see what happens on the buildbots.

Would you mind adding a comment above the blacklist in both ports mentioning that this is for C++20 support? Thanks.

Comment (fracai) [reply devel/resource_dasm/Portfile:11]: Thanks, I had meant to add such a comment, but neglected to add it before pushing.
Latest push addresses the dependency categories as well.

File: devel/resource_dasm/Portfile
Comment (fracai) [reply devel/resource_dasm/Portfile:33]: Thanks. I noted that when building the Portfile and intended to add them under different categories, but simply forgot to get them sorted.

**Review by harens — APPROVED** [review]
Thank you for your contribution @fracai! 🎉

**Pull request merged**
